io-sim/src/Control/Monad/IOSim/Types.hs

```diff
@@ -832,26 +833,25 @@ data StmTxResult s a =
        --
      | StmTxAborted  [SomeTVar s] SomeException
 
+
+-- | OrElse/Catch give rise to an alternate right hand side branch. A right branch
+-- can be a NoOp
+data BranchStmA s a = OrElseStmA (StmA s a) | NoOpStmA
+
 data StmStack s b a where
   -- | Executing in the context of a top level 'atomically'.
   AtomicallyFrame  :: StmStack s a a
 
-  -- | Executing in the context of the /left/ hand side of an 'orElse'
-  OrElseLeftFrame  :: StmA s a                -- orElse right alternative
+  -- | Executing in the context of the /left/ hand side of a branch.
+  -- A right branch is represented by a frame containing empty statement.
+  BranchFrame      :: BranchStmA s a          -- right alternative, can be empty
```

Comment:
It might not matter, but my intuition is that this `BranchStmA` argument should be strict.

Reply:
I agree, it ought to be strict.

Reply:
This is a GADT, is there a way to mark it as strict? The function `execAtomically - go`  works with strict arguments. Also we can take care of it while constructing `BranchFrame` .

Reply:
```suggestion
  BranchFrame      :: !(BranchStmA s a)       -- right alternative, can be empty
```

Reply:
Got it. I was missing parenthesis, and error wasn't very helpful.